conan/tools/qbs/qbsdeps.py

```diff
@@ -0,0 +1,161 @@
+from conan.tools.files import save
+from conan.errors import ConanException
+from conan.tools.env import VirtualBuildEnv
+from conans.model.dependencies import get_transitive_requires
+import json
+import os
+
+
+class _QbsDepsFile:
+    def __init__(self, qbsdeps):
+        self._qbsdeps = qbsdeps
+
+    @property
+    def filename(self):
+        return None
+
+    def get_content(self):
+        return {}
+
+    def render(self):
+        return json.dumps(self.get_content(), indent=4)
+
+
+class _QbsDepsCommonFile(_QbsDepsFile):
+    def __init__(self, qbsdeps, format_version=1):
+        super().__init__(qbsdeps)
+        self._format_version = format_version
+
+    @property
+    def filename(self):
+        return 'common.json'
+
+    def get_content(self):
+        env_vars = VirtualBuildEnv(self._qbsdeps._conanfile).vars()
```

Comment:
```suggestion
        env_vars = VirtualBuildEnv(self._qbsdeps._conanfile, , auto_generate=True).vars()
```
Otherwise, it disable the generation of the user side VirtualBuildEnv
This doesn't look like a ``QbsDeps`` thing, but in any case it should be part of the ``QbsToolchain`` generator, isn't it?

How is this ``common.json`` file used by Qbs?

Reply:
We use the common.json (and environment from it) to find tools for cross-compiled builds - e.g. protobuf module for ARM contains ARM protoc, but we need to use the amd64 version which can be found in PATH in build env. 
So basically we generate a "host" module for linking, but that module also embeds the info about "build" tools - that way we only need one dependency on Qbs side.

I am open for other suggestions though - the PR [0] in Qbs is not merged yet (though it covers use-cases I am aware of) so it's not too late to change the format.

[0] https://codereview.qt-project.org/c/qbs/qbs/+/336750

Reply:
Here's what the resulting module looks like [0]. Since we embed build env in every module - maybe do the same in QbsDeps and omit common.json altogether? I kinda started with one huge file with all deps in a dict, but since I moved away from it, maybe its remnants should go as well. Leads to some duplication thought but maybe it's a minor issue.

[0] https://pastebin.com/kRQJctHp

Reply:
> We use the common.json (and environment from it) to find tools for cross-compiled builds - e.g. protobuf module for ARM contains ARM protoc, but we need to use the amd64 version which can be found in PATH in build env.
So basically we generate a "host" module for linking, but that module also embeds the info about "build" tools - that way we only need one dependency on Qbs side.

At the moment the "build" context dependencies are not being generated. I'd probably suggest starting incrementally, supported by ``functional`` tests, the moment we add dependencies from the build context, we add the build-env, but not earlier.

Reply:
It took me a while to understand what you mean.
Indeed, "build" deps are not generated since there was no immediate need for this. I see a couple of use-cases though and I will probably add those. However, Qbs uses the (merged) env (PATH, namely) from build deps (installed via [tool_require]) in order to locate build tools.

By design, Qbs does not distinguish between "host" and "build" modules explicitly. There might be multiple modules for different platforms/architectures within one project (meaning, more than two aka host and build). Usually, we check if arch/platform/whatever of the module match those in a Product that is currently being built. Thus, a "host" module should also contain info about build tools in order to execute those tools. That's why I insert PATH in every json and later in the module file so that when Qbs loads the "host" module, it can also locate "build" tools.

It just occurred to me that instead of using VirtualEnv and using PATH variable, I can take the location of build tools from the "build" version of the dependency and inject it in the "host" module via a variable, say "build_bindirs".
Hopefully this makes sense. Thoughts?

PS: I'll add a functional test in order to demonstrate how commands should be executed

Reply:
> It just occurred to me that instead of using VirtualEnv and using PATH variable, I can take the location of build tools from the "build" version of the dependency and inject it in the "host" module via a variable, say "build_bindirs".
Hopefully this makes sense. Thoughts?

In general, it is much better to work and add things incrementally. Starting from the most basic use case, in this case would be just libraries in the host context via regular requires. 

That allows maintainers also to get some understand and expertise about how tools such as qbs work, maybe to install it in CI, depending on the complexity. After some experience, likely improving a bit the code or some interfaces/ux, it is easier to add more advanced use cases, like injecting tools from ``tool_requires``, environment, etc.